### PR TITLE
Add `rename()` for DataFrame and LazyFrame

### PR DESCRIPTION
**Pull request by etiennebacher**
Not sure about which args we should have here. I used `existing` and `new` as in the [Rust implementation](https://github.com/pola-rs/polars/blob/814025213d098e66fa73d5f43e14f686c994b282/py-polars/src/lazyframe.rs#L668) but the [Python implementation](https://pola-rs.github.io/polars/py-polars/html/reference/dataframe/api/polars.DataFrame.rename.html#polars.DataFrame.rename) uses a named list like `list(old = "new")` that I also like.

What do you think?

``` r
library(polars)

pl$
  DataFrame(mtcars)$
  rename(c("mpg", "hp"), c("miles_per_gallon", "horsepower"))
#> shape: (32, 11)
#> ┌──────────────────┬─────┬───────┬────────────┬───┬─────┬─────┬──────┬──────┐
#> │ miles_per_gallon ┆ cyl ┆ disp  ┆ horsepower ┆ … ┆ vs  ┆ am  ┆ gear ┆ carb │
#> │ ---              ┆ --- ┆ ---   ┆ ---        ┆   ┆ --- ┆ --- ┆ ---  ┆ ---  │
#> │ f64              ┆ f64 ┆ f64   ┆ f64        ┆   ┆ f64 ┆ f64 ┆ f64  ┆ f64  │
#> ╞══════════════════╪═════╪═══════╪════════════╪═══╪═════╪═════╪══════╪══════╡
#> │ 21.0             ┆ 6.0 ┆ 160.0 ┆ 110.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0  ┆ 4.0  │
#> │ 21.0             ┆ 6.0 ┆ 160.0 ┆ 110.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0  ┆ 4.0  │
#> │ 22.8             ┆ 4.0 ┆ 108.0 ┆ 93.0       ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0  ┆ 1.0  │
#> │ 21.4             ┆ 6.0 ┆ 258.0 ┆ 110.0      ┆ … ┆ 1.0 ┆ 0.0 ┆ 3.0  ┆ 1.0  │
#> │ …                ┆ …   ┆ …     ┆ …          ┆ … ┆ …   ┆ …   ┆ …    ┆ …    │
#> │ 15.8             ┆ 8.0 ┆ 351.0 ┆ 264.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0  ┆ 4.0  │
#> │ 19.7             ┆ 6.0 ┆ 145.0 ┆ 175.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0  ┆ 6.0  │
#> │ 15.0             ┆ 8.0 ┆ 301.0 ┆ 335.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0  ┆ 8.0  │
#> │ 21.4             ┆ 4.0 ┆ 121.0 ┆ 109.0      ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0  ┆ 2.0  │
#> └──────────────────┴─────┴───────┴────────────┴───┴─────┴─────┴──────┴──────┘
```

<sup>Created on 2023-06-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Review by eitsupi — APPROVED** [review]
Awesome! Looks nice.

**Comment by etiennebacher** [comment]
Actually I definitely prefer using a named list, it's much easier to see what is replaced by what when we want to rename a lot of variables:

``` r
library(polars)

pl$
  DataFrame(mtcars)$
  rename(
    list(
      mpg = "miles_per_gallon", 
      hp = "horsepower",
      gear = "Number of forward gears",
      carb = "Number of carburators"
    )
  )
#> shape: (32, 11)
#> ┌────────────────────┬─────┬───────┬────────────┬───┬─────┬─────┬────────────────────┬─────────────┐
#> │ miles_per_gallon   ┆ cyl ┆ disp  ┆ horsepower ┆ … ┆ vs  ┆ am  ┆ Number of forward  ┆ Number of   │
#> │ ---                ┆ --- ┆ ---   ┆ ---        ┆   ┆ --- ┆ --- ┆ gears              ┆ carburators │
#> │ f64                ┆ f64 ┆ f64   ┆ f64        ┆   ┆ f64 ┆ f64 ┆ ---                ┆ ---         │
#> │                    ┆     ┆       ┆            ┆   ┆     ┆     ┆ f64                ┆ f64         │
#> ╞════════════════════╪═════╪═══════╪════════════╪═══╪═════╪═════╪════════════════════╪═════════════╡
#> │ 21.0               ┆ 6.0 ┆ 160.0 ┆ 110.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0                ┆ 4.0         │
#> │ 21.0               ┆ 6.0 ┆ 160.0 ┆ 110.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0                ┆ 4.0         │
#> │ 22.8               ┆ 4.0 ┆ 108.0 ┆ 93.0       ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0                ┆ 1.0         │
#> │ 21.4               ┆ 6.0 ┆ 258.0 ┆ 110.0      ┆ … ┆ 1.0 ┆ 0.0 ┆ 3.0                ┆ 1.0         │
#> │ …                  ┆ …   ┆ …     ┆ …          ┆ … ┆ …   ┆ …   ┆ …                  ┆ …           │
#> │ 15.8               ┆ 8.0 ┆ 351.0 ┆ 264.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 4.0         │
#> │ 19.7               ┆ 6.0 ┆ 145.0 ┆ 175.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 6.0         │
#> │ 15.0               ┆ 8.0 ┆ 301.0 ┆ 335.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 8.0         │
#> │ 21.4               ┆ 4.0 ┆ 121.0 ┆ 109.0      ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0                ┆ 2.0         │
#> └────────────────────┴─────┴───────┴────────────┴───┴─────┴─────┴────────────────────┴─────────────┘
```

<sup>Created on 2023-06-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by eitsupi** [comment]
> Actually I definitely prefer using a named list, it's much easier to see what is replaced by what when we want to rename a lot of variables:

I agree with that, but I think the `old = "new"` style is confusing...
I think `new = "old"` is better.

**Comment by etiennebacher** [comment]
I prefer `old = "new"` but other packages like `dplyr` have this `new = old` so I suppose it's better for consistency to also have `new = old`

**Comment by etiennebacher** [comment]
New syntax:

``` r
library(polars)

pl$
  DataFrame(mtcars)$
  rename(
    list(
      miles_per_gallon = "mpg", 
      horsepower = "hp",
      `Number of forward gears` = "gear",
      `Number of carburators` = "carb"
    )
  )
#> shape: (32, 11)
#> ┌────────────────────┬─────┬───────┬────────────┬───┬─────┬─────┬────────────────────┬─────────────┐
#> │ miles_per_gallon   ┆ cyl ┆ disp  ┆ horsepower ┆ … ┆ vs  ┆ am  ┆ Number of forward  ┆ Number of   │
#> │ ---                ┆ --- ┆ ---   ┆ ---        ┆   ┆ --- ┆ --- ┆ gears              ┆ carburators │
#> │ f64                ┆ f64 ┆ f64   ┆ f64        ┆   ┆ f64 ┆ f64 ┆ ---                ┆ ---         │
#> │                    ┆     ┆       ┆            ┆   ┆     ┆     ┆ f64                ┆ f64         │
#> ╞════════════════════╪═════╪═══════╪════════════╪═══╪═════╪═════╪════════════════════╪═════════════╡
#> │ 21.0               ┆ 6.0 ┆ 160.0 ┆ 110.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0                ┆ 4.0         │
#> │ 21.0               ┆ 6.0 ┆ 160.0 ┆ 110.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0                ┆ 4.0         │
#> │ 22.8               ┆ 4.0 ┆ 108.0 ┆ 93.0       ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0                ┆ 1.0         │
#> │ 21.4               ┆ 6.0 ┆ 258.0 ┆ 110.0      ┆ … ┆ 1.0 ┆ 0.0 ┆ 3.0                ┆ 1.0         │
#> │ …                  ┆ …   ┆ …     ┆ …          ┆ … ┆ …   ┆ …   ┆ …                  ┆ …           │
#> │ 15.8               ┆ 8.0 ┆ 351.0 ┆ 264.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 4.0         │
#> │ 19.7               ┆ 6.0 ┆ 145.0 ┆ 175.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 6.0         │
#> │ 15.0               ┆ 8.0 ┆ 301.0 ┆ 335.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 8.0         │
#> │ 21.4               ┆ 4.0 ┆ 121.0 ┆ 109.0      ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0                ┆ 2.0         │
#> └────────────────────┴─────┴───────┴────────────┴───┴─────┴─────┴────────────────────┴─────────────┘
```

<sup>Created on 2023-06-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/dataframe__frame.R [hunk 1]
```diff
@@ -1366,3 +1366,15 @@ DataFrame_pivot = function(
...
+#' @keywords DataFrame
+#' @param mapping A named list like `list(new = "old")`.
+#' @return DataFrame
+#' @examples
+#' pl$DataFrame(mtcars)$
+#'   lazy()$
+#'   rename(list(miles_per_gallon = "mpg", horsepower = "hp"))
+DataFrame_rename = function(mapping) {
+  self$lazy()$rename(mapping)$collect()
+}
```
Comment (eitsupi) [inline R/dataframe__frame.R:1380]: Would it be possible to omit list() from the argument by changing it like follows?

```suggestion
DataFrame_rename = function(...) {
  self$lazy()$rename(list(...))$collect()
}
```

Comment (etiennebacher) [reply R/dataframe__frame.R:1380]: I've got to go now so I'll take a look later

Comment (etiennebacher) [reply R/dataframe__frame.R]: done

**Comment by etiennebacher** [comment]
``` r
library(polars)

pl$
  DataFrame(mtcars)$
  rename(
    miles_per_gallon = "mpg", 
    horsepower = "hp",
    `Number of forward gears` = "gear",
    `Number of carburators` = "carb"
  )
#> shape: (32, 11)
#> ┌────────────────────┬─────┬───────┬────────────┬───┬─────┬─────┬────────────────────┬─────────────┐
#> │ miles_per_gallon   ┆ cyl ┆ disp  ┆ horsepower ┆ … ┆ vs  ┆ am  ┆ Number of forward  ┆ Number of   │
#> │ ---                ┆ --- ┆ ---   ┆ ---        ┆   ┆ --- ┆ --- ┆ gears              ┆ carburators │
#> │ f64                ┆ f64 ┆ f64   ┆ f64        ┆   ┆ f64 ┆ f64 ┆ ---                ┆ ---         │
#> │                    ┆     ┆       ┆            ┆   ┆     ┆     ┆ f64                ┆ f64         │
#> ╞════════════════════╪═════╪═══════╪════════════╪═══╪═════╪═════╪════════════════════╪═════════════╡
#> │ 21.0               ┆ 6.0 ┆ 160.0 ┆ 110.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0                ┆ 4.0         │
#> │ 21.0               ┆ 6.0 ┆ 160.0 ┆ 110.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 4.0                ┆ 4.0         │
#> │ 22.8               ┆ 4.0 ┆ 108.0 ┆ 93.0       ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0                ┆ 1.0         │
#> │ 21.4               ┆ 6.0 ┆ 258.0 ┆ 110.0      ┆ … ┆ 1.0 ┆ 0.0 ┆ 3.0                ┆ 1.0         │
#> │ …                  ┆ …   ┆ …     ┆ …          ┆ … ┆ …   ┆ …   ┆ …                  ┆ …           │
#> │ 15.8               ┆ 8.0 ┆ 351.0 ┆ 264.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 4.0         │
#> │ 19.7               ┆ 6.0 ┆ 145.0 ┆ 175.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 6.0         │
#> │ 15.0               ┆ 8.0 ┆ 301.0 ┆ 335.0      ┆ … ┆ 0.0 ┆ 1.0 ┆ 5.0                ┆ 8.0         │
#> │ 21.4               ┆ 4.0 ┆ 121.0 ┆ 109.0      ┆ … ┆ 1.0 ┆ 1.0 ┆ 4.0                ┆ 2.0         │
#> └────────────────────┴─────┴───────┴────────────┴───┴─────┴─────┴────────────────────┴─────────────┘
```

<sup>Created on 2023-06-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by etiennebacher** [comment]
@eitsupi I'm just re-requesting your review since there were changes since you approved

**Comment by sorhawell** [comment]
r-polars has in general a syntax deviation from py-polars . Wherever in py-polars where first input is a list or a dict e.g. `pl.DataFrame({"a":[1,2,3]})` then `pl$DataFrame(a=1:3)` and `pl$DataFrame(list(a=1:3))` are both valid. Also e.g. `df.select([pl.col("a")])` then `df$select(pl$col("a"))`. If enable named_expr it is also possible to write `df$select(new_a=pl$col("a"))` instead of $alias("new"). But the ladder has not stabilized in py-polars due to wildcard last I checked.

**Comment by etiennebacher** [comment]
so @sorhawell are you fine with the syntax in the last post?

**Comment by sorhawell** [comment]
> so @sorhawell are you fine with the syntax in the last post?

Oh yes I missed that :) sry! Looks great :)

I will take a quick look at code also

**Review by sorhawell — APPROVED** [review]

**Comment by sorhawell** [comment]
forgot to add something

**Review by eitsupi — APPROVED** [review]
Awesome! Thanks!

**Pull request merged**